---
title: "`r params$report_title`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  report_title: "Calculating codon bias scores"
  report_author: "Tamas Szabo"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  codon_of_interest: "AGA"
  codon_table: "../codon_table.csv"
  freq_folder: "../codon_freqs"
---

## Setup

```{r message = FALSE, warning = FALSE}
# Import tools to be used

library(dplyr)
library(ggplot2)

select <- dplyr::select
filter <- dplyr::filter
```

```{r}
# Codon table to map amino acids

codon_tab <- params$codon_table %>%
  read.csv() %>%
  mutate(
    Codon = gsub("T", "U", Codon)
  )

COI <- params$codon_of_interest
COMPETITORS <- codon_tab %>%
  group_by(Aa) %>%
  mutate(
    INTEREST = COI %in% Codon,
  ) %>%
  filter(INTEREST & Codon != COI) %>%
  .$Codon

XP_COMPETE <- 1 / (length(COMPETITORS) + 1)

print(COMPETITORS)
```

```{r}
# Codon frequencies calculated before

frq_tab <- params$freq_folder %>%
  dir(full.names=TRUE) %>%
  setNames(., basename(.)) %>%
  purrr::map_dfr(read.csv, .id="SPECIES") %>%
  dplyr::rename(TRANSCRIPT = X) %>%
  mutate(
    SPECIES = gsub("_cds_.*", "", SPECIES),
    TRANSCRIPT = gsub(".*_", "", TRANSCRIPT)
  )

distinct(frq_tab, SPECIES, .keep_all=TRUE)
```

## Bias calculation

```{r}
# Calculate score for codon

bias_tab <- frq_tab %>%
  select(one_of(c('SPECIES', 'TRANSCRIPT', COI, COMPETITORS))) %>%
  group_by(SPECIES, TRANSCRIPT) %>%
  summarise_if(is.numeric, mean, na.rm = TRUE) %>%
  ungroup() %>%
  rename(BASE = sym(COI)) %>%
  mutate(
    total = rowSums(across(where(is.numeric))),
    SCORE = ((BASE / total) - XP_COMPETE) /XP_COMPETE
  )

head(bias_tab)
```

```{r}
# Show distribution of score

bias_tab %>%
  ggplot(aes(x=SCORE)) +
  geom_density()
```


