---
title: "`r params$report_title`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  report_title: "Calculating codon bias scores"
  report_author: "Tamas Szabo"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  codon_of_interest: "AGA"
  output_prefix: "../example_outputs/"
  codon_in_prefix: true
  top_n: 250
  codon_table: "../stock_data/codon_table.csv"
  freq_folder: "../stock_data/codon_freqs"
---

## Setup

```{r message = FALSE, warning = FALSE}
# Import tools to be used

library(dplyr)
library(purrr)
library(ggplot2)
library(ggsci)
library(clusterProfiler)
library(msigdbr)
library(AnnotationDbi)
library(org.Sc.sgd.db)
library(org.Dm.eg.db)

select <- dplyr::select
filter <- dplyr::filter

if (params$codon_in_prefix) {
  EXPORT_PREFIX <- paste(params$output_prefix, params$codon_of_interest, sep="")
} else {
  EXPORT_PREFIX <- params$output_prefix
}

export_dir <- dirname(paste(EXPORT_PREFIX, "x.csv", sep="_"))
if (!dir.exists(export_dir)) dir.create(export_dir)
```

```{r}
# Codon table to map amino acids

codon_tab <- params$codon_table %>%
  read.csv() %>%
  mutate(
    Codon = gsub("T", "U", Codon)
  )

COI <- params$codon_of_interest
COMPETITORS <- codon_tab %>%
  group_by(Aa) %>%
  mutate(
    INTEREST = COI %in% Codon,
  ) %>%
  filter(INTEREST & Codon != COI) %>%
  .$Codon

XP_COMPETE <- 1 / (length(COMPETITORS) + 1)

print(COMPETITORS)
```

```{r}
# Codon frequencies calculated before

frq_tab <- params$freq_folder %>%
  dir(full.names=TRUE) %>%
  setNames(., basename(.)) %>%
  purrr::map_dfr(read.csv, .id="SPECIES") %>%
  dplyr::rename(GENEID = X) %>%
  mutate(
    SPECIES = gsub("_cds_.*", "", SPECIES),
    GENEID = gsub(".*_", "", GENEID)
  )

distinct(frq_tab, SPECIES, .keep_all=TRUE)
```

```{r}
# Fix gene IDs for yeast and drosi

sc_sub <- frq_tab %>%
  filter(SPECIES == "yeast")

id_map <- AnnotationDbi::select(org.Sc.sgd.db, keys=sc_sub$GENEID, columns="ENTREZID", keytype="ENSEMBL") %>%
  filter(!is.na(ENTREZID)) %>%
  {setNames(.$ENTREZID, .$ENSEMBL)}

sc_sub <- sc_sub %>%
  mutate(GENEID = id_map[GENEID]) %>%
  filter(!is.na(GENEID))

dr_sub <- frq_tab %>%
  filter(SPECIES == "drosophila")

id_map <- AnnotationDbi::select(org.Dm.eg.db, keys=dr_sub$GENEID, columns="ENTREZID", keytype="FLYBASE") %>%
  filter(!is.na(ENTREZID)) %>%
  {setNames(.$ENTREZID, .$FLYBASE)}

dr_sub <- dr_sub %>%
  mutate(GENEID = id_map[GENEID]) %>%
  filter(!is.na(GENEID))

frq_tab <- frq_tab %>%
  filter(!SPECIES %in% c("drosophila", "yeast")) %>%
  bind_rows(dr_sub, sc_sub)

distinct(frq_tab, SPECIES, .keep_all=TRUE)
```

## Bias calculation

```{r}
# Calculate score for codon

coi_setter <- function(x) {
  x$BASE <- x[[COI]]
  x
}

bias_tab <- frq_tab %>%
  select(one_of(c('SPECIES', 'GENEID', COI, COMPETITORS))) %>%
  group_by(SPECIES, GENEID) %>%
  summarise_if(is.numeric, mean, na.rm=TRUE) %>%
  ungroup() %>%
  coi_setter %>%
  mutate(
    total = rowSums(across(where(is.numeric))),
    SCORE = ((BASE / total) - XP_COMPETE) * XP_COMPETE
  )

head(bias_tab)
```

## Score stats

### Top scoring genes

```{r}
top_tags <- bias_tab %>%
  arrange(desc(SCORE)) %>%
  select(SPECIES, GENEID, SCORE) %>%
  group_by(SPECIES) %>%
  group_split() %>%
  map_dfr(head, params$top_n)

write.csv(top_tags, paste(EXPORT_PREFIX, "toptags.csv", sep="_"), row.names=FALSE)
```

### Score distribution

```{r}
# Show distribution of score

p <- bias_tab %>%
  mutate(
    SPECIES = factor(SPECIES, levels=c("human", "mouse", "zebrafish", "drosophila", "yeast"))
  ) %>%
  ggplot(aes(x=SCORE, color=SPECIES)) +
  geom_density(fill=NA) +
  scale_color_npg() +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  labs(x="Bias score", title=paste("Codon usage bias of", COI))

ggsave(paste(EXPORT_PREFIX, "score_distribution.pdf", sep="_"), p, width=4, height=3)

p
```

## Biassed pathways

```{r}
# Retrieve MSigDB collections for model species

msig_list <- list(
  human = "Homo sapiens",
  mouse = "Mus musculus",
  zebrafish = "Danio rerio",
  drosophila = "Drosophila melanogaster",
  yeast = "Saccharomyces cerevisiae"
) %>%
  lapply(function(x) msigdbr(species=x, category="C5", subcategory="BP")[,c('gs_name', 'entrez_gene')])
```

### ORA

```{r}
top_tags <- bias_tab %>%
  arrange(desc(SCORE)) %>%
  select(SPECIES, GENEID, SCORE) %>%
  group_by(SPECIES) %>%
  group_split() %>%
  map_dfr(function(x) {
    sp_l <- unique(x$SPECIES)
    x %>%
      filter(GENEID %in% msig_list[[sp_l]]$entrez_gene) %>%
      head(params$top_n)
  })
```

```{r}
enrichment <- msig_list %>%
  purrr::imap(~clusterProfiler::enricher(
    top_tags[top_tags$SPECIES == .y,]$GENEID,
    TERM2GENE=.x,
    pAdjustMethod="none")
  )
```

```{r}
plot_data <- enrichment %>%
  map_dfr(function(x) x@result, .id="SPECIES") %>%
  mutate(
    Description = gsub("GO_", "", Description),
    Description = gsub("_", " ", Description),
    setSize = strsplit(GeneRatio, "/"),
    setSize = map(setSize, function(x) x[[2]])
  ) %>%
  unnest(setSize) %>%
  arrange(p.adjust) %>%
  group_by(SPECIES) %>%
  mutate(
    pw_rank = 50 - seq(0, n()-1)
  ) %>%
  ungroup() %>%
  group_by(ID) %>%
  mutate(
    pw_sum_rank = sum(pw_rank, na.rm=TRUE)
  ) %>%
  ungroup() %>%
  select(Description, SPECIES, Count, setSize, pvalue, p.adjust, pw_rank, pw_sum_rank) %>%
  arrange(desc(pw_sum_rank))

head(plot_data)
```

```{r}
top_pathways <- plot_data %>%
  .$Description %>%
  unique() %>%
  head(15)

p <- plot_data %>%
  filter(Description %in% top_pathways) %>%
  mutate(
    logPadj = -log10(p.adjust),
    Description = factor(Description, levels=rev(top_pathways)),
    SPECIES = factor(SPECIES, levels=names(msig_list))
  ) %>%
  ggplot(aes(x=SPECIES, y=Description, size=logPadj, color=SPECIES)) +
  geom_point() +
  scale_color_npg() +
  theme_bw() +
  theme(
    axis.text.x=element_text(angle=30, hjust=1),
    axis.text.y=element_text(size=8)
  ) +
  labs(
    x = "", y = "", title = paste("Top biassed pathways for codon", params$codon_of_interest)
  )

ggsave(paste(EXPORT_PREFIX, "ora_dotplot.pdf", sep="_"), p, width=9, height=6)

p
```

### GSEA

```{r}
enriched_sets <- msig_list %>%
  purrr::imap(~clusterProfiler::GSEA(
    bias_tab %>% filter(SPECIES == .y & !is.na(SCORE)) %>% arrange(desc(SCORE)) %>% {setNames(.$SCORE, .$GENEID)},
    TERM2GENE=.x, pAdjustMethod="none")
  )
```

```{r}
plot_data <- enriched_sets %>%
  map_dfr(function(x) x@result, .id="SPECIES") %>%
  mutate(
    Description = gsub("GO_", "", Description),
    Description = gsub("_", " ", Description),
    Count = strsplit(core_enrichment, "/"),
    Count = map(Count, length)
  ) %>%
  arrange(desc(abs(NES))) %>%
  group_by(SPECIES) %>%
  mutate(
    pw_rank = 50 - seq(0, n()-1)
  ) %>%
  ungroup() %>%
  group_by(ID) %>%
  mutate(
    pw_sum_rank = sum(pw_rank, na.rm=TRUE)
  ) %>%
  ungroup() %>%
  select(Description, SPECIES, Count, setSize, NES, p.adjust, pw_rank, pw_sum_rank) %>%
  arrange(desc(pw_sum_rank))

head(plot_data)
```

```{r}
top_pathways <- plot_data %>%
  .$Description %>%
  unique() %>%
  head(15)

p <- plot_data %>%
  filter(Description %in% top_pathways) %>%
  mutate(
    logPadj = -log10(p.adjust),
    Description = factor(Description, levels=rev(top_pathways)),
    SPECIES = factor(SPECIES, levels=names(msig_list))
  ) %>%
  ggplot(aes(x=SPECIES, y=Description, size=logPadj, color=NES)) +
  geom_point() +
  scale_color_gradientn(
    colors=rev(c('#173beb', '#4dbbd5ff', '#bee7f7', 'grey', 'white', 'grey', '#f7b09c', '#e64b35ff', '#ff0000')),
    breaks=c(-3, -2, -2, -1, 0, 1, 2, 2, 3), limits=c(-3.5, 3.5), oob = scales::squish
  ) +
  theme_bw() +
  theme(
    axis.text.x=element_text(angle=30, hjust=1),
    axis.text.y=element_text(size=8)
  ) +
  labs(
    x = "", y = "", title = paste("Top biassed pathways for codon", params$codon_of_interest)
  )

ggsave(paste(EXPORT_PREFIX, "gsea_dotplot.pdf", sep="_"), p, width=9, height=6)

p
```