---
title: "`r params$report_title`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  report_title: "Calculating codon bias scores"
  report_author: "Tamas Szabo"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  codon_of_interest: "AGA"
  codon_table: "../codon_table.csv"
  freq_folder: "../codon_freqs"
---

## Setup

```{r message = FALSE, warning = FALSE}
# Import tools to be used

library(dplyr)
library(purrr)
library(ggplot2)
library(ggsci)
library(clusterProfiler)
library(msigdbr)
library(AnnotationDbi)
library(org.Sc.sgd.db)
library(org.Dm.eg.db)

select <- dplyr::select
filter <- dplyr::filter
```

```{r}
# Codon table to map amino acids

codon_tab <- params$codon_table %>%
  read.csv() %>%
  mutate(
    Codon = gsub("T", "U", Codon)
  )

COI <- params$codon_of_interest
COMPETITORS <- codon_tab %>%
  group_by(Aa) %>%
  mutate(
    INTEREST = COI %in% Codon,
  ) %>%
  filter(INTEREST & Codon != COI) %>%
  .$Codon

XP_COMPETE <- 1 / (length(COMPETITORS) + 1)

print(COMPETITORS)
```

```{r}
# Codon frequencies calculated before

frq_tab <- params$freq_folder %>%
  dir(full.names=TRUE) %>%
  setNames(., basename(.)) %>%
  purrr::map_dfr(read.csv, .id="SPECIES") %>%
  dplyr::rename(TRANSCRIPT = X) %>%
  mutate(
    SPECIES = gsub("_cds_.*", "", SPECIES),
    TRANSCRIPT = gsub(".*_", "", TRANSCRIPT)
  )

distinct(frq_tab, SPECIES, .keep_all=TRUE)
```

```{r}
# Fix gene IDs for yeast and drosi

sc_sub <- frq_tab %>%
  filter(SPECIES == "yeast")

id_map <- AnnotationDbi::select(org.Sc.sgd.db, keys=sc_sub$TRANSCRIPT, columns="ENTREZID", keytype="ENSEMBL") %>%
  filter(!is.na(ENTREZID)) %>%
  {setNames(.$ENTREZID, .$ENSEMBL)}

sc_sub <- sc_sub %>%
  mutate(TRANSCRIPT = id_map[TRANSCRIPT]) %>%
  filter(!is.na(TRANSCRIPT))

dr_sub <- frq_tab %>%
  filter(SPECIES == "drosophila")

id_map <- AnnotationDbi::select(org.Dm.eg.db, keys=dr_sub$TRANSCRIPT, columns="ENTREZID", keytype="FLYBASE") %>%
  filter(!is.na(ENTREZID)) %>%
  {setNames(.$ENTREZID, .$FLYBASE)}

dr_sub <- dr_sub %>%
  mutate(TRANSCRIPT = id_map[TRANSCRIPT]) %>%
  filter(!is.na(TRANSCRIPT))

frq_tab <- frq_tab %>%
  filter(!SPECIES %in% c("drosophila", "yeast")) %>%
  bind_rows(dr_sub, sc_sub)

distinct(frq_tab, SPECIES, .keep_all=TRUE)
```


## Bias calculation

```{r}
# Calculate score for codon

coi_setter <- function(x) {
  x$BASE <- x[[COI]]
  x
}

bias_tab <- frq_tab %>%
  select(one_of(c('SPECIES', 'TRANSCRIPT', COI, COMPETITORS))) %>%
  group_by(SPECIES, TRANSCRIPT) %>%
  summarise_if(is.numeric, mean, na.rm=TRUE) %>%
  ungroup() %>%
  coi_setter %>%
  mutate(
    total = rowSums(across(where(is.numeric))),
    SCORE = ((BASE / total) - XP_COMPETE) * XP_COMPETE
  )

head(bias_tab)
```

```{r}
# Show distribution of score

bias_tab %>%
  ggplot(aes(x=SCORE, fill=SPECIES)) +
  geom_density() +
  scale_fill_npg(alpha=0.6) +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  labs(x="Bias score", title=paste("Codon usage bias of", COI))
```

```{r}
# Show distribution of score

bias_tab %>%
  ggplot(aes(x=SCORE, color=SPECIES)) +
  geom_density(fill=NA) +
  scale_color_npg() +
  theme_bw() +
  theme(
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank()
  ) +
  labs(x="Bias score", title=paste("Codon usage bias of", COI))
```

## Top biassed proteins

```{r}
TOP_N <- 150

top_tags <- bias_tab %>%
  arrange(desc(SCORE)) %>%
  select(SPECIES, TRANSCRIPT, SCORE) %>%
  group_split(SPECIES) %>%
  map(head, TOP_N) %>%
  bind_rows()

top_tags
```

## Biassed pathways

```{r}
# Retrieve MSigDB collections for model species

msig_list <- list(
  human = "Homo sapiens",
  mouse = "Mus musculus",
  zebrafish = "Danio rerio",
  drosophila = "Drosophila melanogaster",
  yeast = "Saccharomyces cerevisiae"
) %>%
  lapply(function(x) msigdbr(species=x, category="C5", subcategory="BP")[,c('gs_name', 'entrez_gene')])
```

### ORA

```{r}
enrichment <- msig_list %>%
  purrr::imap(~clusterProfiler::enricher(
    top_tags[top_tags$SPECIES == .y, "TRANSCRIPT"],
    TERM2GENE=.x
    pAdjustMethod="none")
  )
```

