---
title: "`r params$report_title`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  report_title: "Parsing CDS sequences from a genome"
  report_author: "Tamas Szabo"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  species: "human"
  fasta_folder: "../fasta_transcriptomes"
---

## Setup

```{r message = FALSE, warning = FALSE}
# Import tools to be used

library(dplyr)
library(GenomicFeatures)

select <- dplyr::select
filter <- dplyr::filter
```

```{r}
# Genomes and annotations for model organisms

library(BSgenome.Hsapiens.UCSC.hg19)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(BSgenome.Mmusculus.UCSC.mm10)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(BSgenome.Drerio.UCSC.danRer10)
library(TxDb.Drerio.UCSC.danRer10.refGene)
library(BSgenome.Dmelanogaster.UCSC.dm6)
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)
library(BSgenome.Scerevisiae.UCSC.sacCer3)
library(TxDb.Scerevisiae.UCSC.sacCer3.sgdGene)
```

```{r}
# Create output folder if it does not exist

if(!dir.exists(params$fasta_folder)) dir.create(params$fasta_folder, recursive=TRUE)
```

## Whole genome CDSs

## Homo sapiens

```{r message = FALSE, warning = FALSE}
tx_set <- TxDb.Hsapiens.UCSC.hg19.knownGene
genome_seqs <- BSgenome.Hsapiens.UCSC.hg19

cdsBy(tx_set, "tx") %>%
  extractTranscriptSeqs(genome_seqs, .) %>%
  RNAStringSet() %>%
  writeXStringSet(file.path(params$fasta_folder, "human_cds.fa"))
```

## Mus musculus

```{r message = FALSE, warning = FALSE}
tx_set <- TxDb.Mmusculus.UCSC.mm10.knownGene
genome_seqs <- BSgenome.Mmusculus.UCSC.mm10

cdsBy(tx_set, "tx") %>%
  extractTranscriptSeqs(genome_seqs, .) %>%
  RNAStringSet() %>%
  writeXStringSet(file.path(params$fasta_folder, "mouse_cds.fa"))
```

## Danio rerio

```{r message = FALSE, warning = FALSE}
tx_set <- TxDb.Drerio.UCSC.danRer10.refGene
genome_seqs <- BSgenome.Drerio.UCSC.danRer10

cdsBy(tx_set, "tx") %>%
  extractTranscriptSeqs(genome_seqs, .) %>%
  RNAStringSet() %>%
  writeXStringSet(file.path(params$fasta_folder, "zebrafish_cds.fa"))
```

## Drosophila melanogaster

```{r message = FALSE, warning = FALSE}
tx_set <- TxDb.Dmelanogaster.UCSC.dm6.ensGene
genome_seqs <- BSgenome.Dmelanogaster.UCSC.dm6

cdsBy(tx_set, "tx") %>%
  extractTranscriptSeqs(genome_seqs, .) %>%
  RNAStringSet() %>%
  writeXStringSet(file.path(params$fasta_folder, "drosophila_cds.fa"))
```

## Yeast

```{r message = FALSE, warning = FALSE}
tx_set <- TxDb.Scerevisiae.UCSC.sacCer3.sgdGene
genome_seqs <- BSgenome.Scerevisiae.UCSC.sacCer3

cdsBy(tx_set, "tx") %>%
  extractTranscriptSeqs(genome_seqs, .) %>%
  RNAStringSet() %>%
  writeXStringSet(file.path(params$fasta_folder, "yeast_cds.fa"))
```

```{r}
frq_tab <- "../codon_freqs" %>%
  dir(full.names=TRUE) %>%
  setNames(., basename(.)) %>%
  purrr::map_dfr(read.csv, .id="SPECIES") %>%
  dplyr::rename(TRANSCRIPT = X) %>%
  mutate(
    SPECIES = gsub("_cds_.*", "", SPECIES)
  )

head(frq_tab)
```

