---
title: "`r params$report_title`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  report_title: "Parsing CDS sequences from a genome"
  report_author: "Tamas Szabo"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  species: "human"
  fasta_folder: "../fasta_transcriptomes"
---

## Setup

```{r message = FALSE, warning = FALSE}
# Import tools to be used

library(dplyr)
library(GenomicFeatures)

select <- dplyr::select
filter <- dplyr::filter
```
```{r}
if(!dir.exists(params$fasta_folder)) dir.create(params$fasta_folder, recursive=TRUE)
```

```{r}
BiocManager::install('TxDb.Dmelanogaster.UCSC.dm6.ensGene')
BiocManager::install('TxDb.Drerio.UCSC.danRer10.refGene')
```


```{r}
library(EnsDb.Hsapiens.v86)
library(EnsDb.Mmusculus.v79)
library(BSgenome.Drerio.UCSC.danRer10)
library(BSgenome.Dmelanogaster.UCSC.dm6)
library(TxDb.Drerio.UCSC.danRer10.refGene)
library(TxDb.Dmelanogaster.UCSC.dm6.ensGene)

hsdb <- EnsDb.Hsapiens.v86
mmdb <- EnsDb.Mmusculus.v79
drdb <- BSgenome.Drerio.UCSC.danRer10
dmdb <- BSgenome.Dmelanogaster.UCSC.dm6

hs_dna <- ensembldb:::getGenomeTwoBitFile(hsdb)
mm_dna <- ensembldb:::getGenomeTwoBitFile(mmdb)
```


```{r message = FALSE, warning = FALSE}
tx_set <- TxDb.Dmelanogaster.UCSC.dm6.ensGene
genome_seqs <- BSgenome.Dmelanogaster.UCSC.dm6

cdsBy(tx_set, "tx") %>%
  extractTranscriptSeqs(genome_seqs, .) %>%
  RNAStringSet() %>%
  writeXStringSet(file.path(params$fasta_folder, "drosophila_cds.fa"))
```

