---
title: "`r params$report_title`"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: united
    code_folding: hide
params:
  report_title: "Pathways affected by codon biassed genes"
  report_author: "Tamas Szabo"
  report_time: !r format(Sys.Date(), format="%d/%m/%Y")
  species: "human"
  freq_folder: "../codon_freqs"
---

## Setup

```{r message = FALSE, warning = FALSE}
# Import tools to be used

library(dplyr)
library(ggplot2)
library(clusterProfiler)
library(msigdbr)

select <- dplyr::select
filter <- dplyr::filter
```

```{r}
# Mock object

mydf <- data.frame(Entrez=c('1', '100', '1000', '100101467','100127206', '100128071'), group = c('A', 'A', 'A', 'B', 'B', 'B'))
emptyRes <- compareCluster(Entrez~group, data=mydf, fun="enrichGO", 'org.Hs.eg.db')
```

## GO pathway collections

```{r}
hs_geneSet <- "Homo sapiens" %>%
  msigdbr(species=., category="C5", subcategory="BP") %>%
  .[,c('gs_name', 'entrez_gene')]
```

```{r}
mm_geneSet <- "Mus musculus" %>%
  msigdbr(species=., category="C5", subcategory="BP") %>%
  .[,c('gs_name', 'entrez_gene')]
```

```{r}
dr_geneSet <- "Danio rerio" %>%
  msigdbr(species=., category="C5", subcategory="BP") %>%
  .[,c('gs_name', 'entrez_gene')]
```

```{r}
dm_geneSet <- "Drosophila melanogaster" %>%
  msigdbr(species=., category="C5", subcategory="BP") %>%
  .[,c('gs_name', 'entrez_gene')]
```

```{r}
sc_geneSet <- "Saccharomyces cerevisiae" %>%
  msigdbr(species=., category="C5", subcategory="BP") %>%
  .[,c('gs_name', 'entrez_gene')]
```

```{r}
msig_list <- list(
  human = hs_geneSet,
  mouse = mm_geneSet,
  zebrafish = dr_geneSet,
  drosophils = dm_geneSet,
  yeast = sc_geneSet
)
```

### Codon freqs

```{r}
frq_tab <- params$freq_folder %>%
  dir(full.names=TRUE) %>%
  setNames(., basename(.)) %>%
  purrr::map_dfr(read.csv, .id="SPECIES") %>%
  dplyr::rename(TRANSCRIPT = X) %>%
  mutate(
    SPECIES = gsub("_cds_.*", "", SPECIES),
    TRANSCRIPT = gsub(".*_", "", TRANSCRIPT)
  )

distinct(frq_tab, SPECIES, .keep_all=TRUE)
```

```{r}
sc_sub <- frq_tab %>%
  filter(SPECIES == "yeast")

id_map <- AnnotationDbi::select(org.Sc.sgd.db, keys=sc_sub$TRANSCRIPT, columns="ENTREZID", keytype="ENSEMBL") %>%
  filter(!is.na(ENTREZID)) %>%
  {setNames(.$ENTREZID, .$ENSEMBL)}

sc_sub <- sc_sub %>%
  mutate(TRANSCRIPT = id_map[TRANSCRIPT]) %>%
  filter(!is.na(TRANSCRIPT))

head(sc_sub)
```

```{r}
dr_sub <- frq_tab %>%
  filter(SPECIES == "drosophila")

id_map <- AnnotationDbi::select(org.Dm.eg.db, keys=dr_sub$TRANSCRIPT, columns="ENTREZID", keytype="FLYBASE") %>%
  filter(!is.na(ENTREZID)) %>%
  {setNames(.$ENTREZID, .$FLYBASE)}

dr_sub <- dr_sub %>%
  mutate(TRANSCRIPT = id_map[TRANSCRIPT]) %>%
  filter(!is.na(TRANSCRIPT))

head(sc_sub)
```

```{r}
frq_tab <- frq_tab %>%
  filter(!SPECIES %in% c("drosophila", "yeast")) %>%
  bind_rows(dr_sub, sc_sub)
```

### Hitlist bias

```{r}
richRes <- data.frame(
  Cluster=c(),
  group=c(),
  ID=c(),
  Description=c(),
  GeneRatio=c(),
  BgRatio=c(),
  pvalue=c(),
  p.adjust=c(),
  qvalue=c(),
  geneID=c(),
  Count=c()
  )

compNames <- list()
i <- 0
for (compname in names(msig_list)){
  geneSet <- msig_list[[compname]]
  hitGenIt <- frq_tab %>%
    filter(SPECIES == compname) %>%
    select(TRANSCRIPT, GGA, CCA, AAA, SUM) %>%
    mutate(WOBBLE = (GGA + CCA + AAA) / SUM) %>%
    group_by(TRANSCRIPT) %>%
    mutate(WOBBLE = mean(WOBBLE)) %>%
    ungroup() %>%
    distinct(TRANSCRIPT, .keep_all=TRUE) %>%
    arrange(WOBBLE) %>%
    tail(250) %>%
    .$TRANSCRIPT
  
  i <- i + 1
  #compname <- paste0(compname, ' (', length(hitGenIt), ')')
  compNames[[i]] <- compname
  enrichment <- clusterProfiler::enricher(hitGenIt, TERM2GENE=geneSet, pAdjustMethod="none")
  if(!is.null(enrichment)){
    erik <- head(enrichment@result, n=20)
    erik$Cluster <- rep(compname, nrow(erik))
    erik$group <- rep(compname, nrow(erik))
    richRes <- rbind(erik, richRes)
  }
}
richRes <- richRes[order(richRes$p.adjust), ]
richRes$Count <- (as.numeric(unlist(strsplit(richRes$GeneRatio, '/'))) / as.numeric(unlist(strsplit(richRes$BgRatio, '/'))))[seq(1, length(richRes$GeneRatio)*2, 2)]
richRes$Count <- richRes$Count * 100
compRes <- rlang::duplicate(emptyRes)
richRes$group <- factor(richRes$group, levels=compNames)
richRes$Cluster <- factor(richRes$Cluster, levels=compNames)
compRes@compareClusterResult <- richRes
```

```{r}
multi_hitlist_enrichdot <- function(enrichment, plot_title=""){
  resulTab <- enrichment@compareClusterResult %>%
    mutate(logPAD = -log10(p.adjust)) %>%
    arrange(p.adjust) %>%
    group_by(group) %>%
    mutate(ConditionRank = seq(1, n())) %>%
    ungroup() %>%
    mutate(ConditionRank = 30 - ConditionRank) %>%
    group_by(ID) %>%
    mutate(
      sumP = sum(logPAD, na.rm=TRUE),
      sumRank = sum(ConditionRank, na.rm=TRUE)
    ) %>%
    ungroup() %>%
    arrange(desc(sumRank), desc(sumP))
  
  topsets <- unique(resulTab$ID)[1:50]
  hitnames <- levels(resulTab$group)
  resulTab %>%
    ggplot(aes(x=group, y=ID, size=Count, color=p.adjust)) +
    geom_point() +
    labs(title=plot_title) +
    scale_color_gradientn(
      colors=rev(c( '#2b8cbe', '#2b8cbe', 'grey', '#e38071', '#e34a33', '#e31e00')),
      breaks=c(0.05, 0.01, 0.001, 0.0001, 0.00001),
      limits=c(0.000001, 1), trans='log10', oob = scales::squish
    ) +
    scale_y_discrete(name="", limits=rev(topsets), labels=rev(topsets)) +
    scale_x_discrete(name="", labels=hitnames) +
    theme(
      axis.text.x=element_text(angle=30, hjust=1),
      axis.text.y=element_text(size=8)
    )
  
}

p <- multi_hitlist_enrichdot(compRes)
p
```

